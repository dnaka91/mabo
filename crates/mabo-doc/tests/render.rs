use std::{
    fmt::Write,
    fs,
    path::{Path, PathBuf},
};

use insta::{assert_snapshot, glob, with_settings};
use mabo_doc::{Opts, Output};
use mabo_parser::Schema;

fn strip_path(path: &Path) -> PathBuf {
    path.strip_prefix(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/inputs"))
        .unwrap()
        .to_owned()
}

fn merge_output(buf: &mut String, output: &Output<'_>, parent: &Path) {
    let path = parent.join(output.name);
    let file = parent.join(&output.file);
    let _ = write!(buf, "\n\n--- {}\n\n{}", file.display(), output.content);

    for module in &output.modules {
        merge_output(buf, module, &path);
    }
}

#[test]
fn render_schema() {
    glob!("inputs/*.mabo", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str(), Some(&strip_path(path))).unwrap();
        let value = mabo_compiler::simplify_schema(&value);
        let value = mabo_doc::render_schema(&Opts {}, &value).unwrap();

        let mut merged = String::new();
        merge_output(&mut merged, &value, Path::new(""));

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("render", merged.trim_start());
        });
    });
}
