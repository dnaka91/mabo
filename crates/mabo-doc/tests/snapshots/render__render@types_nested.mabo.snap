---
source: crates/mabo-doc/tests/render.rs
description: "struct Sample {\n    value: Vec<Option<NonZero<HashMap<i64, Box<String>>>>> @1,\n}"
input_file: crates/mabo-parser/tests/inputs/types_nested.mabo
snapshot_kind: text
---
--- types_nested/index.html

<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />

    
    <link rel="stylesheet" href="../assets/style.css" />

    <title>types_nested - Mabo</title>
  </head>
  <body>
    <nav>
    </nav>
    <main>
<div class="section">
  <h2>Schema types_nested</h2>
  <div class="markdown pl-6">
    
  </div>
</div>
<div class="section">
  <h3>Modules</h3>
  <table>
  </table>
</div>
<div class="section">
  <h3>Structs</h3>
  <table>
    <tr>
      <td class="pr-8">
        <a class="text-lg name-struct" href="struct.Sample.html">Sample</a>
      </td>
      <td>
        
      </td>
    </tr>
  </table>
</div>
<div class="section">
  <h3>Enums</h3>
  <table>
  </table>
</div>
<div class="section">
  <h3>Aliases</h3>
  <table>
  </table>
</div>
<div class="section">
  <h3>Constants</h3>
  <table>
  </table>
</div>
    </main>
  </body>
</html>

--- types_nested/struct.Sample.html

<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />

    
    <link rel="stylesheet" href="../assets/style.css" />

    <title>Sample - Mabo</title>
  </head>
  <body>
    <nav>
    </nav>
    <main>
<div class="section">
  <h2>
    Struct
    <a href="index.html">types_nested</a>::<span class="name-struct">Sample</span>
  </h2>
  <pre class="item-definition my-2">struct Sample {
    value: Vec&#60;Option&#60;NonZero&#60;HashMap&#60;i64, Box&#60;String&#62;&#62;&#62;&#62;&#62; @1,
}</pre>
  <div class="markdown pl-6">
    
  </div>
</div>
<div class="section">
  <h3>Metadata</h3>
  <div class="pl-6 mt-2">
    <p>The next free ID is <code class="font-bold">2</code></p>
  </div>
</div>
<div class="section">
  <h3>Fields</h3>
  <ul>
    <li class="field-definition">
      <code>
        <span class="field-name">value</span>:
        <span class="field-type">Vec&#60;Option&#60;NonZero&#60;HashMap&#60;i64, Box&#60;String&#62;&#62;&#62;&#62;&#62;</span>
        <span class="field-id">@1</span>
      </code>
      <div class="markdown pl-4">
        
      </div>
      <div class="pl-4">
        <h4 class="my-2">Metadata</h4>
        <div class="text-sm">
          <p>The size range is:</p>
          <div
            class="markdown markdown-sm markdown-p:mb-1 markdown-ul:m-0 markdown-ul:pl-4 markdown-li:m-0 markdown-li:pl-0">
            <p><strong>Vec</strong> <code>1..</code></p>
          <ul>
          <li>length: <strong>u64</strong> <code>1..10</code></li>
          <li>element: <strong>Option</strong> <code>0..</code>
          <ul>
          <li>value: <strong>NonZero</strong> <code>0..</code>
          <ul>
          <li>value: <strong>HashMap</strong> <code>1..</code>
          <ul>
          <li>length: <strong>u64</strong> <code>1..10</code></li>
          <li>key: <strong>i64</strong> <code>1..10</code></li>
          <li>value: <strong>Box&lt;String&gt;</strong> <code>1..</code></li>
          </ul>
          </li>
          </ul>
          </li>
          </ul>
          </li>
          </ul>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>

    </main>
  </body>
</html>
